File: backend/app/schemas/quiz.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.schemas.base import BaseSchema, Score, TimestampedDTO, VersionedDTO
from app.schemas.curriculum import CurriculumTopicDTO
from app.schemas.enums import (
    DifficultyLevel,
    QuestionType,
    QuizAttemptStatus,
    QuizStatus,
    ScoringPolicyType,
)
from app.schemas.ids import AttemptId, ConceptId, CurriculumId, GoalId, QuestionId, QuizId, TopicId


class QuizQuestionDTO(BaseSchema):
    question_id: QuestionId
    question_type: QuestionType
    prompt: str = Field(min_length=1, max_length=800)
    concept_ids: list[ConceptId] = Field(min_length=1, max_length=8)
    difficulty: DifficultyLevel
    correct_answer: str = Field(min_length=1, max_length=500)
    points: float = Field(gt=0.0, le=20.0)
    options: list[str] = Field(default_factory=list, max_length=8)
    rubric: str | None = Field(default=None, max_length=1000)
    explanation: str | None = Field(default=None, max_length=1000)


class LearnerQuizQuestionDTO(BaseSchema):
    question_id: QuestionId
    question_type: QuestionType
    prompt: str = Field(min_length=1, max_length=800)
    concept_ids: list[ConceptId] = Field(min_length=1, max_length=8)
    difficulty: DifficultyLevel
    points: float = Field(gt=0.0, le=20.0)
    options: list[str] = Field(default_factory=list, max_length=8)


class QuizScoringPolicy(BaseSchema):
    type: ScoringPolicyType
    partial_credit: bool = False


class QuizDTO(TimestampedDTO, VersionedDTO):
    quiz_id: QuizId
    goal_id: GoalId
    curriculum_id: CurriculumId
    target_topic_ids: list[TopicId] = Field(min_length=1, max_length=20)
    target_concept_ids: list[ConceptId] = Field(min_length=1, max_length=20)
    status: QuizStatus
    title: str = Field(min_length=1, max_length=220)
    questions: list[QuizQuestionDTO] = Field(min_length=1, max_length=20)
    scoring_policy: QuizScoringPolicy
    difficulty: DifficultyLevel | None = None


class LearnerQuizDTO(BaseSchema):
    quiz_id: QuizId
    title: str = Field(min_length=1, max_length=220)
    questions: list[LearnerQuizQuestionDTO] = Field(min_length=1, max_length=20)
    scoring_policy: QuizScoringPolicy


class QuizAnswerSubmission(BaseSchema):
    question_id: QuestionId
    answer_text: str | None = Field(default=None, max_length=1200)
    selected_options: list[str] = Field(default_factory=list, max_length=8)


class ConceptQuizScore(BaseSchema):
    concept_id: ConceptId
    score: Score
    correct_count: int = Field(ge=0, le=100)
    total_questions: int = Field(ge=1, le=100)


class QuizAttemptDTO(TimestampedDTO, VersionedDTO):
    quiz_attempt_id: AttemptId
    quiz_id: QuizId
    goal_id: GoalId
    curriculum_id: CurriculumId
    answers: list[QuizAnswerSubmission] = Field(min_length=1, max_length=50)
    total_score: Score
    correct_count: int = Field(ge=0, le=100)
    total_questions: int = Field(ge=1, le=100)
    concept_scores: list[ConceptQuizScore] = Field(min_length=1)
    weak_concepts: list[ConceptId] = Field(default_factory=list, max_length=20)
    submitted_at: datetime
    status: QuizAttemptStatus = QuizAttemptStatus.SCORED
    feedback: str | None = Field(default=None, max_length=1000)
    adaptation_triggered: bool = False


class QuizAgentInput(BaseSchema):
    goal_text: str = Field(min_length=5, max_length=500)
    curriculum_topics: list[CurriculumTopicDTO] = Field(min_length=1)
    # No min_length: a learner with no known weak concepts yet (fresh
    # workspace, no progress history) legitimately has none to target - the
    # deterministic agent's own concept-prioritization already falls back to
    # the curriculum's own topic concepts when this is empty.
    target_concepts: list[ConceptId] = Field(default_factory=list, max_length=20)
    difficulty: DifficultyLevel
    question_count: int = Field(ge=1, le=20)


class QuizAgentOutput(BaseSchema):
    quiz_title: str = Field(min_length=1, max_length=220)
    questions: list[QuizQuestionDTO] = Field(min_length=1, max_length=20)
    scoring_policy: QuizScoringPolicy


class QuizScoreOutput(BaseSchema):
    total_score: Score
    correct_count: int = Field(ge=0, le=100)
    total_questions: int = Field(ge=1, le=100)
    concept_scores: list[ConceptQuizScore] = Field(min_length=1)
    weak_concepts: list[ConceptId] = Field(default_factory=list, max_length=20)
    feedback: str = Field(min_length=1, max_length=1000)

File: frontend/lib/types/quiz.ts
import type { DifficultyLevel } from "@/lib/types/curriculum";

export type QuestionType = "multiple_choice" | "short_answer" | "self_rating";

export type QuizAttemptStatus = "submitted" | "scored" | "failed";

export type QuizQuestionDTO = {
  question_id: string;
  question_type: QuestionType;
  prompt: string;
  concept_ids: string[];
  difficulty: DifficultyLevel;
  correct_answer: string;
  points: number;
  options: string[];
  rubric: string | null;
  explanation: string | null;
};

export type LearnerQuizQuestionDTO = {
  question_id: string;
  question_type: QuestionType;
  prompt: string;
  concept_ids: string[];
  difficulty: DifficultyLevel;
  points: number;
  options: string[];
};

// GET /quizzes/{quiz_id} - the learner-safe question set (no answer key),
// used to drive the take flow before any attempt exists.
export type LearnerQuizDTO = {
  quiz_id: string;
  title: string;
  questions: LearnerQuizQuestionDTO[];
  scoring_policy: { type: string; partial_credit: boolean };
};

export type QuizAnswerSubmission = {
  question_id: string;
  answer_text: string | null;
  selected_options: string[];
};

export type ConceptQuizScore = {
  concept_id: string;
  score: number;
  correct_count: number;
  total_questions: number;
};

export type QuizAttemptDTO = {
  quiz_attempt_id: string;
  quiz_id: string;
  goal_id: string;
  curriculum_id: string;
  answers: QuizAnswerSubmission[];
  total_score: number;
  correct_count: number;
  total_questions: number;
  concept_scores: ConceptQuizScore[];
  weak_concepts: string[];
  submitted_at: string;
  status: QuizAttemptStatus;
  feedback: string | null;
  adaptation_triggered: boolean;
  created_at: string;
  updated_at: string;
  schema_version: string;
};

// The backend composes `questions` from a plain `QuizQuestionDTO[]` (answer-
// keyed) only once `attempt.status === "scored"`, else from a
// `LearnerQuizQuestionDTO[]` (no answer key) - see
// `backend/app/schemas/quiz.py`'s `QuizAttemptReviewDTO`. `questions`'
// element type is a plain union rather than two full DTO variants of this
// type: the discriminant (`attempt.status`) lives one level down, so
// `isScoredQuizAttemptReview` below is the reliable way to narrow it - a
// direct `review.attempt.status === "scored"` check doesn't reliably narrow
// a nested discriminant for TS.
export type QuizAttemptReviewDTO = {
  attempt: QuizAttemptDTO;
  questions: QuizQuestionDTO[] | LearnerQuizQuestionDTO[];
};

export type ScoredQuizAttemptReviewDTO = {
  attempt: QuizAttemptDTO & { status: "scored" };
  questions: QuizQuestionDTO[];
};

export function isScoredQuizAttemptReview(
  review: QuizAttemptReviewDTO
): review is ScoredQuizAttemptReviewDTO {
  return review.attempt.status === "scored";
}
